### ISSUE TITLE
[enhancement] Biome Selector

### ISSUE THREAD

**Issue by Webber88** [DarkholmeTenk/TardisModIssues#440]
much like the lever for demensions maybe add a lever that would lock on to the nearest selected biome(i thought of this since im having a lot of trouble finding the biomes i need due to all the mods i use)

**Comment by AlienXtream** [comment]
i was kinda wanting something like this but i never suggested it because i figured it would require more work than its worth

**Comment by FoxPotato** [comment]
I think this was already requested once, It wouldn't be too hard i think? just an if() in a forloop. And then check if it's the requested biome. But i'd have to check if it's possible to check what biome a block is in. (should be, how else is foliage coloured)

**Comment by AlienXtream** [comment]
if this were implemented then getting biome from external scanners is a must

**Comment by DarkholmeTenk** [comment]
This has already been suggested. Working out the biome you are in is easy, but finding a biome is hard.

**Comment by FoxPotato** [comment]
Am I still allowed to try though? Hard != Impossible :P

**Comment by AlienXtream** [comment]
well if you did a grid search in an area centered on the tardis's location (10 000 blocks?) and say checked every 100 blocks for what the biome would be then i COULD be doable. not sure how taxing that would be on system resources though. thats why the programing is left to you 2 and i just break it :P

**Comment by FoxPotato** [comment]
That sound doable, that's only a 100 by 100 square. just 2 nestled for loops.
But maybe the steps should be bigger, i mean. Biomes are quite big right? I'd say steps of 250 would do the trick.

**Comment by Webber88** [comment]
:3

**Comment by AlienXtream** [comment]
well it should be configurable because some mods (like biome o plenty) have a config option that lets you make biomes smaller. though why you would have smaller biomes is beyond me but that should be accounted for no less

**Comment by FoxPotato** [comment]
Yeah, dark told me it should be a config option. So that's what's gonna happen :P

**Comment by AlienXtream** [comment]
sweat, more things for me to break :P \o/

**Comment by FoxPotato** [comment]
I was thinking of having like 1 button and 1 lever. Lever to select the biome. And button to set the controls to the nearest point that the loop can find. Thoughts?
I can't really think of another way to do this :P

**Comment by AlienXtream** [comment]
that could work, though if not too much extra effort should also have a "lazy" search mode that dosn't look for a specific biome but just the temperature/type i.e. snowy, mountainous, ocean ect. the "lazy" mode being unlocked earlier perhaps? though the name may be a little ironic from a performance perspective as for that insted of just matching a string you have to compare an array of strings right? i think im starting to understand the logistics a bit better now.

**Comment by AlienXtream** [comment]
nah, just go with K.I.S.S button and lever. but insted of an actual LEVER could you use a "wheel"? i just feel like it would fit better

**Comment by FoxPotato** [comment]
I guess it should be possible to check what kind of biome it is. But i don't know how useful that would be.

**Comment by FoxPotato** [comment]
And biomes get assigned an enum, so performance wise it wouldn't be too heavy. :P

**Comment by AlienXtream** [comment]
ok well like i said K.I.S.S. lets just see how messy the selector gets with large biome mods (like Biomes o plenty) and if its worth having that "lazy" mode.

**Comment by FoxPotato** [comment]
I don't actually know what you mean by K.I.S.S. D:

**Comment by AlienXtream** [comment]
keep it simple stupid. its just something that ppl say when things are getting to complex :p

**Comment by FoxPotato** [comment]
ahhh :P 
Well, it's not that hard. Could do a config option so people can choose if they want specific biomes. or the types.

```
    MESA,
    FOREST,
    PLAINS,
    MOUNTAIN,
    HILLS,
    SWAMP,
    SANDY,
    SNOWY,
    WASTELAND,
    BEACH,
```

or these generic types.

**Comment by FoxPotato** [comment]
Also, there doesn't seem to be any real room anymore on the console. And seeing as it's a control it wouldn't really fit on the engine :/

**Comment by AlienXtream** [comment]
well the coordinate guesser is practically redundant since the screen came out so you COULD replace that

**Comment by FoxPotato** [comment]
I'll ask dark if he's ok with that. Good thinking :P 

EDIT: actually, would that be smart? next to the delete all rooms buttons? It's a 2 press thing, but still.

**Comment by AlienXtream** [comment]
tbh ive kinda been hoping that thing would get some love since the screen came out

**Comment by AlienXtream** [comment]
good point :/

**Comment by AlienXtream** [comment]
it could also go just above the helmic regulator and have the biome display on the screen

**Comment by FoxPotato** [comment]
it'd have to be quite small. and the button wouldnt fit, i kinda want those 2 together. else it'll be such a mess

**Comment by AlienXtream** [comment]
another option is to move the room deletion button to the engine. that way its also out of the way and practically impossible to accidentally activate

**Comment by FoxPotato** [comment]
It would actually fit better there, yeah.
First I'm trying to find a way to have it not lag when looking for the biome tho, quite hard to get it right with nestled loops :P

**Comment by AlienXtream** [comment]
have it run slower on by design. delay the loop to allow time for the search, then in each step print a % to chat to indicate progress. unless my understanding of things is wrong, in which case just ignore the extraterestrial :P

**Comment by FoxPotato** [comment]
As you probably know code runs from bottom to top from left to right. When it finds a loop, it'll pause the current thread until the loop is done. So the server won't update until the loop is done. Hence creating server lagg.

**Comment by AlienXtream** [comment]
bottom to top?! all i know of coding is the base logic (like the if, then, else logic :P) so yeh... could you break the loop earlier and have it do sequential loops instead of nested ones?

**Comment by FoxPotato** [comment]
I stop both loops as soon as one set of coordinates within that biome has been found. but if that's like near the end of the loops, you'll still get some lagg. Maybe if i just tweak the amount of blocks it skips every time it looks. and maybe the size of the grid i can minimize it as much as possible.
